### ISSUE TITLE
[UX] Add configuration to exclude node types from search indexing

### ISSUE THREAD

**Issue by alanmels** [backdrop/backdrop-issues#4197]
**Description of the need**

Here is the excerpt from https://github.com/backdrop-contrib/search_index_options:

> The core search module scans all the available content, regardless of whether some content types were meant to be searchable at all or not. This kind of design causes extra processing load while indexing and storing lots of unnecessary information in database. 

**Proposed solution**

The quote goes on as:

> The Search Index Options module addresses this problem by letting users to set only desired content types for indexing, thus significantly reducing the load and amount of stored index data.

I suggest to save Backdrop from repeating one of the most illogical features Drupal has - it takes all the nodes into database, and only after that the modules start interacting and operating with the search index data. Why can't we let users to choose which content types or nodes they want to be indexed?! If you look at the module's code it's pretty simple and straightforward. So I suggest to move some features of the module to core. 

**Additional information**

Here is the stats screenshot of a test Backdrop site with 52 nodes in total, 24 of which are of **Page** type (indexed) and rest of **Post** type (not indexed). 

<img width="877" alt="Screen Shot 2019-11-06 at 1 42 36 AM" src="https://user-images.githubusercontent.com/56604405/68252610-04aefe80-0037-11ea-9c74-2095971c6e22.png">

With Post content type not indexed at all, we store only 46% of initial data in the database. Remember those huge index tables of Drupal sites? Now, we can do it right way on Backdrop core!

---

~PR by @alanmels https://github.com/backdrop/backdrop/pull/2990~
PR by @quicksketch https://github.com/backdrop/backdrop/pull/3129 (based on @alanmels's PR)

**Comment by jenlampton** [comment]
Thanks @alanmels I was about to open the same issue! 
Here's the relevant Drupal issue: https://www.drupal.org/node/111744

**Comment by alanmels** [comment]
Thanks, Jen. Yes, the issue page on DO references the Search Index Options module. Knowing it could take forever to get such a feature on Drupal core, we coded a simple module and contributed to DO. 

Now on Backdrop hopefully we can quickly get it to core. I could take a stab and create a PR if you take this idea positively.

**Comment by jenlampton** [comment]
> Now on Backdrop hopefully we can quickly get it to core. I could take a stab and create a PR if you take this idea positively.

Yes please! :)

**Comment by herbdool** [comment]
I think this is a duplicate. I don't have time to look for the issue but there might already be a PR to deal with rabbit hole functionality, if I recall correctly.

**Comment by herbdool** [comment]
There was this already merged https://github.com/backdrop/backdrop/pull/2824 that addresses rabbit hole. Though it doesn't touch indexing, it just changes access: people with right permission should still be able to see the rabbit hole results in search.

So this issue isn't exactly the same but still might be a duplicate. I recall something like it already.

**Comment by alanmels** [comment]
@herbdool, thanks for the link. I've looked at it and as I understand it is about permissions for paths. Whereas, the feature I am requesting here directly deals with search index table in the database. The goal is to significantly reduce the search index tables size. So two are really different issues.

**Comment by jenlampton** [comment]
> I think this is a duplicate. 

@herbdool I also thought there might have been a similar issue. I looked pretty hard for one but couldn't find it. 

Thanks for the reference to the rabbit hole PR! That puts my mind at ease a little :)

**Comment by indigoxela** [comment]
> Remember those huge index tables of Drupal sites? Now, we can do it right way on Backdrop core!

I fully agree on that! Allowing site admins to completely exclude types from indexing would be a real improvement.

@alanmels thanks for sharing this module. I agree, this functionality should make its way into core. :+1:

**Comment by klonos** [comment]
Another upvote from me here. Huge databases are causing all sorts of issues in the Lagoon/OpenShift-based distribution cluster the company I work for is providing support for. We should do things right when it makes sense, and this one does 👍 

I would suggest that the default should be to include all content types in the index, and allow site admins to tweak/exclude as needed. The only thing that we could/should perhaps automate is to exclude rabbit-hole'd content types from the index (or alternatively provide an option for disabling index if the rabbit hole option is enabled).

**Comment by alanmels** [comment]
We will work on this issue soon, hopefully on one of the coming days.

**Comment by alanmels** [comment]
Unlike Search Index Options module (which has several different new features including per node indexing), the PR on https://github.com/backdrop/backdrop/pull/2990 introduces only per content type indexing configuration options.

**Comment by indigoxela** [comment]
I was about to begin testing, but realized that many tests are failing with this PR.

@alanmels could you dig into this?

**Comment by alanmels** [comment]
@indigoxela, we have slightly higher workload these days on AltaGrade. However, please be assured someone from our team will sit on this in one of coming days.

**Comment by indigoxela** [comment]
@alanmels that's OK, take your time. I posted just in case you missed the failing tests. You didn't, obviously. ;-)

**Comment by alanmels** [comment]
To reference relevant discussions on Gitter [November 25, 2019 6:10 PM](https://gitter.im/backdrop/backdrop-issues?at=5ddbeef81bf5192e66a15a02):

> @alanmels: Can you take a look at fails messages on https://zen.ci/backdrop/backdrop/test/test-php53-backdrop_backdrop_2990-3560e57-1 and give any idea how to resolve them?
> 
> @klonos: @alanmels yeah, I believe that the most likely reason why you have been getting those errors is because search_install() may be getting called earlier in the setup, when no content types have been created yet. If my assumption is correct, then any statement within the foreach would never be executed; so in that case you'd end up with an undefined $content_types.
> 
> @alanmels: @klonos, if that’s the case then is there any way to make search module install after the node module? Maybe something like https://www.drupal.org/docs/7/creating-custom-modules/howtos/how-to-update-a-modules-weight?
> 
> @klonos: @alanmels I would need someone with better knowledge of the internals to verify my theory, and then yes, we should see that something like that happens, so that we don't bump into this problem.
> ...if I have things right in my head, normally, config that needs to be added when a module is enabled needs to go into a /config folder, in .json files. But in this case, the settings need to be dynamically generated (different content type names). So the install hook is the way to go.

**Comment by docwilmot** [comment]
Node types are created in the install profile, which runs last, so you wont be able to have Search module run after it.
I would recommend you hard code 'page' and 'post' in the Search module config for new installs and add an update hook (using same code as you just tried in the `install` file) for upgrades.

**Comment by alanmels** [comment]
> Node types are created in the install profile, which runs last, so you wont be able to have Search module run after it.
> I would recommend you hard code 'page' and 'post' in the Search module config for new installs and add an update hook (using same code as you just tried in the `install` file) for upgrades.

@docwilmot, thanks for the suggestion. I also initially wanted to use the hardcoded approach, but was reluctant believing there would be objections. Now that you've confirmed node types are created last I believe we have no other options but to hardcode two default content types as enabled by default options. I'll change the PR later today.

**Comment by alanmels** [comment]
Finally, the [PR](https://github.com/backdrop/backdrop/pull/2990) has passed all the checks. Please start reviewing.

**Comment by alanmels** [comment]
Just incase if you wonder why bunch of fieldsets on Search config page (`admin/config/search/settings`) disappear when the Content module is unselected, please note that all the options under the Content Ranking, Indexing fieldsets have effect only when nodes are searchable. They do not have any effect over when only users are searchable.

**Comment by docwilmot** [comment]
Havent reviewed code, but if this is porting functionality from https://github.com/backdrop-contrib/search_index_options, then it would be good to include an upgrade path from that module as well, as well as for persons directly upgrading from D7.

Incidentally noted https://github.com/backdrop-contrib/search_index_options doesnt maintain the D7 commit history but thats a different issue

And would also need new tests for the new functionality before it gets in core.

**Comment by alanmels** [comment]
@docwilmot thanks for the comments. The https://github.com/backdrop-contrib/search_index_options deals with mostly the same UI settings, so if accepted then users of the contributed module for both Backdrop and Drupal 7 would just find the same features on UI. I mean I don't know what should be such an upgrade path about.

As for commit credits for D7 please read https://github.com/backdrop-contrib/search_index_options/issues/2

**Comment by docwilmot** [comment]
Sites with `Search index options` already installed should be able to preserve their settings when they upgrade their Backdrop version after this is committed without uninstalling it first; and sites converting from Drupal should have the same thing. And this PR could also seek to disable `Search index options` on upgrade, especially if function names are shared.

**Comment by alanmels** [comment]
@docwilmot, now I see what you mean. I agree with each of your ideas and will try to update the PR. Thanks a lot!

**Comment by alanmels** [comment]
Revisiting this issue I have to say that:

- I don't believe there are lot's of Backdrop sites with `Search index options` already installed. In fact stats on https://github.com/backdrop-contrib/search_index_options/pulse show none. 

- I do believe sites converting from Drupal will want to re-index their content. I don't think they should keep using the same search index on the new CMS. And be assured that out of [183 Drupal websites](https://www.drupal.org/project/usage/search_index) that use the module the absolute majority are our (AltaGrade) customers. And when we start migrating them all to Backdrop we will enforce indexing their content on Backdrop from a fresh.

So I hope we won't complicate things without much necessity and start pushing issues forward. Please run more tests and review this again as we won't

**Comment by docwilmot** [comment]
I'm adding this to a milestone as it was already a candidate. Will review later, unless someone else does.

**Comment by quicksketch** [comment]
I posted a code review: https://github.com/backdrop/backdrop/pull/2990#pullrequestreview-400531992

**Comment by quicksketch** [comment]
Alan noted he was going to be unavailable for a few weeks, I'll pick this on up to reroll it with the feedback from myself and @docwilmot.

**Comment by quicksketch** [comment]
New PR up at https://github.com/backdrop/backdrop/pull/3129. I made some adjustments:

- I removed the expanded messaging (10 of 200 page nodes indexed, etc) because it requires changes to `hook_search_status()`. The previous PR just stopped using it entirely.
- In addition to preventing disabled node types from being indexed, I also disabled them entirely from search results.
- Moved code around to have node module own all it's own search integration instead of being split between search and node.
- Fixed the update hook and made it so that if no content types are selected, it's the same as all types being selected. That makes the initial install easier since both page and post should be included in search out of box.

We need test coverage on the new functionality to check that excluded types are not included in the index or searched.

**Comment by herbdool** [comment]
I can confirm it works as expected. I've tested selecting/unselecting different content types. Content will appear/disappear in the search.

If have some time I'll look into writing a test for this.

**Comment by quicksketch** [comment]
I adapted some of our existing tests to cover these new node type options.

I also realized/found that adding "node_types" to `search.settings.json` is both unnecessary and possibly an encumbrance to people updating. Since search module doesn't include node settings in the default config, it being absent is going to be common for sites that have never saved the search settings form.

PR: https://github.com/backdrop/backdrop/pull/3129

**Comment by jenlampton** [comment]
Thoughts:
* THIS IS FANTASTIC. This is the #1 request I get from my clients about how search works, and before this it was way harder than it needed to be. I always ended up googling a Lullabot article from 10 years ago on how to alter search query results...
* The form API field label should probably be `Enabled content types` instead of `Enabled types` even though it is in a `Content settings` section. Other than that, the UI is intuitive and works exactly as expected (with the one exception that things are not indexed as fast as I'd like...  but that's another issue all together!)

**Comment by jenlampton** [comment]
In the code review I found one instance of of an adverb missing its `ly` and fixed that in the error message below (fix committed to pr)

`Please select at least one content type to search, or complete disable searching content.` Became `Please select at least one content type to search, or completely disable searching content.`

Otherwise, it looks great! There are lots of other tiny improvements to the search settings form in there too :) 

RTBC!

**Comment by quicksketch** [comment]
Thanks for fixing those grammatical issues.

Thanks @herbdool, @docwilmot, @klonos, and @jenlampton for your reviews.

I've merged https://github.com/backdrop/backdrop/pull/3129 into 1.x for 1.16.0. Thanks @alanmels for both filing this issue and the PR!

**Comment by alanmels** [comment]
Despite unfortunately I can't participate these days I am really glad this has finally been implemented. I the way Backdrop is developing. Be all safe till better days!

**Comment by klonos** [comment]
Thanks for pushing that along @quicksketch and @jenlampton ...I was working on this on my local, but never managed to complete it and push changes due to life happening 😓 ...I'll see if there's anything worth including from my changes and file a follow-up if so.

Thanks again 👍

**Comment by jenlampton** [comment]
I had a few sites that were running search_index_options, and after a core update to 1.16.0 the sites started experiencing some serious issues. 

Is it too late to add an update hook to disable the `search_index_options` module?
